{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/maayanlab/Documents/projects/human-melanoma-signature\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/maayanlab/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.\n",
      "  from ._conv import register_converters as _register_converters\n"
     ]
    }
   ],
   "source": [
    "%cd /Users/maayanlab/Documents/projects/human-melanoma-signature\n",
    "import sys\n",
    "sys.path.append('pipeline/scripts/')\n",
    "import pandas as pd\n",
    "import h5py\n",
    "import numpy as np\n",
    "import json, operator, math, requests\n",
    "from geode_jupies import *\n",
    "import sklearn.metrics.pairwise as smp\n",
    "from scipy.spatial.distance import cosine\n",
    "import seaborn as sns; sns.set(color_codes=True)\n",
    "import RNAseq as RNA\n",
    "from collections import Counter\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/maayanlab/Documents/projects/human-melanoma-signature'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Melanoma import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Code Development for Melanoma Pipeline\n",
    "## Define input and output files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# infiles S4.4\n",
    "#infile = 's4-enrichment.dir/melanoma-enrichr-results.txt'\n",
    "#infile2 = 's4-enrichment.dir/enrichr-results-top.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# infiles\n",
    "infile1 = 's2-signatures.dir/melanoma-intersection-2000.txt'\n",
    "infile2 = 's2-signatures.dir/melanoma-intersection-3000.txt'\n",
    "infile3 = 's2-signatures.dir/melanoma-intersection-5000.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# infiles S2.2\n",
    "sig = 's2-signatures.dir/melanoma-signatures.txt'\n",
    "outfile = 's2-signatures.dir/melanoma-top100-union.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #infiles\n",
    "# cpm = 's1-expression_data.dir/melanoma-cpm.txt'\n",
    "# matches = 's2-expression_data.dir/melanoma-idmatches.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Infile & outfile\n",
    "# infiles = ['s1-expression_data.dir/melanoma-cpm.txt']\n",
    "# outfile = 's1-expression_data.dir/melanoma-IDstatus.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Infile & outfile\n",
    "# infiles = ['rawdata/human_matrix.h5', 'rawdata/m2.csv']\n",
    "# outfile = 's1-expression_data.dir/melanoma-cpm.txt'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "top2000_df = pd.read_table(infile1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = top2000_df['gene'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df=run_enrichr(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_df = pd.read_table(sig).set_index('gene')\n",
    "# Make each signature to a np.array\n",
    "s17a=sig_df['17_a'].values\n",
    "s17b=sig_df['17_b'].values\n",
    "s17c=sig_df['17_c'].values\n",
    "s2a=sig_df['2_a'].values\n",
    "s2b=sig_df['2_b'].values\n",
    "s2c=sig_df['2_c'].values\n",
    "s15a=sig_df['15_a'].values\n",
    "s15b=sig_df['15_b'].values\n",
    "s15c=sig_df['15_c'].values\n",
    "s18a=sig_df['18_a'].values\n",
    "s18b=sig_df['18_b'].values\n",
    "s18c=sig_df['18_c'].values\n",
    "s19a=sig_df['19_a'].values\n",
    "s19b=sig_df['19_b'].values\n",
    "s19c=sig_df['19_c'].values\n",
    "s19d=sig_df['19_d'].values\n",
    "s1=sig_df['1'].values\n",
    "s5a=sig_df['5_a'].values\n",
    "s5b=sig_df['5_b'].values\n",
    "s5c=sig_df['5_c'].values\n",
    "s11=sig_df['11'].values\n",
    "s3=sig_df['3'].values\n",
    "# Take vector sum per cluster as a representation of intersection\n",
    "cluster1_inter= s17a + s17b + s17c + s2a + s2b + s2c\n",
    "cluster2_inter= s15c + s19b + s19d\n",
    "cluster3_inter= s15a + s15b + s3 + s18b + s11\n",
    "cluster4_inter= s18c + s5b +s5c\n",
    "cluster5_inter= s1 + s19c\n",
    "cluster6_inter= s18a + s5a\n",
    "# Sort the array in reserve order\n",
    "sort_idx_1 = np.argsort(cluster1_inter)[::-1]\n",
    "sort_idx_2 = np.argsort(cluster2_inter)[::-1]\n",
    "sort_idx_3 = np.argsort(cluster3_inter)[::-1]\n",
    "sort_idx_4 = np.argsort(cluster4_inter)[::-1]\n",
    "sort_idx_5 = np.argsort(cluster5_inter)[::-1]\n",
    "sort_idx_6 = np.argsort(cluster6_inter)[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['EEF1A1', 'FN1', 'GAPDH', 'ACTB', 'TIMP3', 'ENO1', 'SERPINE2', 'A2M',\n",
       "       'PKM', 'EEF2', 'FLNA', 'HSPA8', 'TTYH3', 'TUBA1B', 'LDHA', 'HSP90AA1',\n",
       "       'VIM', 'AC010970_2', 'RPLP0', 'ACTG1', 'MFSD12', 'TNC', 'SERPINA3',\n",
       "       'HSP90AB1', 'TUBB', 'MYOF', 'UBC', 'SPARC', 'EIF4A1', 'MYH9', 'YWHAZ',\n",
       "       'RPL13A', 'FKBP1A', 'EIF4G2', 'ACTN4', 'RPSA', 'NPM1', 'DCBLD2',\n",
       "       'EIF5A', 'RPL41', 'MT-ND2', 'HNRNPA1', 'SLC16A6', 'ALDOA', 'RACK1',\n",
       "       'RPL4', 'SLC7A5', 'RPS2', 'PPIA', 'YBX1', 'ATP5B', 'RPL15', 'PRNP',\n",
       "       'HDLBP', 'P4HB', 'HSPD1', 'CALU', 'CYR61', 'PFN1', 'ACTN1', 'TGFBI',\n",
       "       'RPS18', 'RPL8', 'RPS19', 'HNRNPA2B1', 'RPS3', 'RPL7', 'ATP2A2', 'LMNA',\n",
       "       'NCL', 'EEF1G', 'RTN4', 'RPL11', 'RPS5', 'RPL7A', 'TUBA1C', 'PABPC1',\n",
       "       'RPL18A', 'RPL6', 'IFI6', 'SEC61A1', 'RPL10', 'CFL1', 'CTNNA1', 'CD109',\n",
       "       'HMGA1', 'S100A6', 'LGALS1', 'NQO1', 'CALR', 'GNAS', 'TPI1', 'PLEC',\n",
       "       'MT-CO3', 'RPL27A', 'SLC3A2', 'RPS4X', 'SLC7A8', 'ANXA2', 'RPS27A'],\n",
       "      dtype='object', name='gene')"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_genes_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Return an index of top 5000 summed genes for this cluster\n",
    "sorted_genes_1 = sig_df.index[sort_idx_1][:50]\n",
    "sorted_genes_2 = sig_df.index[sort_idx_2][:50]\n",
    "sorted_genes_3 = sig_df.index[sort_idx_3][:50]\n",
    "sorted_genes_4 = sig_df.index[sort_idx_4][:50]\n",
    "sorted_genes_5 = sig_df.index[sort_idx_5][:50]\n",
    "sorted_genes_6 = sig_df.index[sort_idx_6][:50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Intersection of all 6 clusters's top 500 differentially exp. genes\n",
    "export=pd.DataFrame (sorted_genes_1 | sorted_genes_2 | sorted_genes_3 | sorted_genes_4 | sorted_genes_5 | sorted_genes_6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 230 entries, 0 to 229\n",
      "Data columns (total 1 columns):\n",
      "gene    230 non-null object\n",
      "dtypes: object(1)\n",
      "memory usage: 1.9+ KB\n"
     ]
    }
   ],
   "source": [
    "export.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "export.to_csv(outfile, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = export['gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df = run_enrichr(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term_name</th>\n",
       "      <th>zscore</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>FDR</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>overlapping_genes</th>\n",
       "      <th>gene_set_library</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein targeting to ER (GO:0045047)</td>\n",
       "      <td>-2.085357</td>\n",
       "      <td>91.031606</td>\n",
       "      <td>5.322236e-17</td>\n",
       "      <td>1.101143e-19</td>\n",
       "      <td>[RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRP-dependent cotranslational protein targetin...</td>\n",
       "      <td>-1.840410</td>\n",
       "      <td>83.694450</td>\n",
       "      <td>1.773965e-17</td>\n",
       "      <td>1.778411e-20</td>\n",
       "      <td>[RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nuclear-transcribed mRNA catabolic process, no...</td>\n",
       "      <td>-1.997175</td>\n",
       "      <td>81.207360</td>\n",
       "      <td>7.292814e-16</td>\n",
       "      <td>2.193327e-18</td>\n",
       "      <td>[RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>extracellular matrix organization (GO:0030198)</td>\n",
       "      <td>-1.641114</td>\n",
       "      <td>71.324472</td>\n",
       "      <td>5.322236e-17</td>\n",
       "      <td>1.333894e-19</td>\n",
       "      <td>[ITGB1, SPARC, COL11A1, SERPINE1, TNC, LAMC1, ...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>peptide biosynthetic process (GO:0043043)</td>\n",
       "      <td>-1.914538</td>\n",
       "      <td>70.881497</td>\n",
       "      <td>1.848915e-14</td>\n",
       "      <td>8.340968e-17</td>\n",
       "      <td>[RPL4, PCSK1, RPL41, RPL3, RPL10, RPS8, RPL23,...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>viral transcription (GO:0019083)</td>\n",
       "      <td>-1.783768</td>\n",
       "      <td>66.887449</td>\n",
       "      <td>1.293451e-14</td>\n",
       "      <td>5.186770e-17</td>\n",
       "      <td>[RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>cellular protein metabolic process (GO:0044267)</td>\n",
       "      <td>-1.179625</td>\n",
       "      <td>65.464684</td>\n",
       "      <td>1.578545e-21</td>\n",
       "      <td>7.912506e-25</td>\n",
       "      <td>[RPL4, RPL3, RPL10, RPLP1, RPLP0, TNC, LAMC1, ...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>regulated exocytosis (GO:0045055)</td>\n",
       "      <td>-2.282059</td>\n",
       "      <td>62.150413</td>\n",
       "      <td>2.118757e-10</td>\n",
       "      <td>1.486847e-12</td>\n",
       "      <td>[SERPINA3, SPARC, VWF, APLP2, SERPINE1, FN1, A...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>cotranslational protein targeting to membrane ...</td>\n",
       "      <td>-1.368407</td>\n",
       "      <td>60.951054</td>\n",
       "      <td>3.010519e-17</td>\n",
       "      <td>4.527096e-20</td>\n",
       "      <td>[RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>viral gene expression (GO:0019080)</td>\n",
       "      <td>-1.577981</td>\n",
       "      <td>59.994607</td>\n",
       "      <td>8.770762e-15</td>\n",
       "      <td>3.077461e-17</td>\n",
       "      <td>[RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>viral process (GO:0016032)</td>\n",
       "      <td>-1.683366</td>\n",
       "      <td>53.959228</td>\n",
       "      <td>2.175298e-12</td>\n",
       "      <td>1.199413e-14</td>\n",
       "      <td>[RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>rRNA metabolic process (GO:0016072)</td>\n",
       "      <td>-1.922755</td>\n",
       "      <td>51.707043</td>\n",
       "      <td>2.411505e-10</td>\n",
       "      <td>2.093540e-12</td>\n",
       "      <td>[RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ncRNA processing (GO:0034470)</td>\n",
       "      <td>-1.905479</td>\n",
       "      <td>51.169028</td>\n",
       "      <td>2.411505e-10</td>\n",
       "      <td>2.175794e-12</td>\n",
       "      <td>[RPL4, SMAD1, RPL41, RPL3, RPL10, RPS8, RPL23,...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>platelet degranulation (GO:0002576)</td>\n",
       "      <td>-1.425928</td>\n",
       "      <td>43.017045</td>\n",
       "      <td>1.214263e-11</td>\n",
       "      <td>7.912493e-14</td>\n",
       "      <td>[SERPINA3, SPARC, VWF, APLP2, SERPINE1, FN1, A...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>rRNA processing (GO:0006364)</td>\n",
       "      <td>-1.601239</td>\n",
       "      <td>42.777656</td>\n",
       "      <td>2.623370e-10</td>\n",
       "      <td>2.498448e-12</td>\n",
       "      <td>[RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>nuclear-transcribed mRNA catabolic process (GO...</td>\n",
       "      <td>-1.227593</td>\n",
       "      <td>42.256861</td>\n",
       "      <td>2.240899e-13</td>\n",
       "      <td>1.123257e-15</td>\n",
       "      <td>[RPL4, EIF4A1, RPL41, RPL3, RPL10, RPS8, RPL23...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>cellular macromolecule biosynthetic process (G...</td>\n",
       "      <td>-1.736787</td>\n",
       "      <td>41.183399</td>\n",
       "      <td>5.020480e-09</td>\n",
       "      <td>5.033063e-11</td>\n",
       "      <td>[RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>neutrophil degranulation (GO:0043312)</td>\n",
       "      <td>-1.981217</td>\n",
       "      <td>39.952382</td>\n",
       "      <td>1.515016e-07</td>\n",
       "      <td>1.746634e-09</td>\n",
       "      <td>[HSPA8, SERPINA3, HSP90AA1, HSP90AB1, ANXA2, T...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>antigen processing and presentation of exogeno...</td>\n",
       "      <td>-2.294106</td>\n",
       "      <td>38.742782</td>\n",
       "      <td>2.980108e-06</td>\n",
       "      <td>4.630745e-08</td>\n",
       "      <td>[HLA-B, HLA-C, HLA-A, B2M, HLA-E]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>regulation of apoptotic process (GO:0042981)</td>\n",
       "      <td>-1.759138</td>\n",
       "      <td>38.454118</td>\n",
       "      <td>3.049322e-08</td>\n",
       "      <td>3.209813e-10</td>\n",
       "      <td>[ITGB1, RPL10, CITED2, MAGED1, CLU, THBS1, FAM...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>neutrophil mediated immunity (GO:0002446)</td>\n",
       "      <td>-1.907256</td>\n",
       "      <td>37.839541</td>\n",
       "      <td>1.856361e-07</td>\n",
       "      <td>2.419317e-09</td>\n",
       "      <td>[HSPA8, SERPINA3, HSP90AA1, HSP90AB1, ANXA2, T...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>translation (GO:0006412)</td>\n",
       "      <td>-1.219466</td>\n",
       "      <td>37.747814</td>\n",
       "      <td>5.990301e-12</td>\n",
       "      <td>3.603189e-14</td>\n",
       "      <td>[RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>antigen processing and presentation of peptide...</td>\n",
       "      <td>-1.728818</td>\n",
       "      <td>35.926213</td>\n",
       "      <td>8.561212e-08</td>\n",
       "      <td>9.440935e-10</td>\n",
       "      <td>[PDIA3, CANX, HLA-B, HLA-C, CALR, HLA-A, B2M, ...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>cytokine-mediated signaling pathway (GO:0019221)</td>\n",
       "      <td>-1.331809</td>\n",
       "      <td>35.830235</td>\n",
       "      <td>2.411505e-10</td>\n",
       "      <td>2.070087e-12</td>\n",
       "      <td>[ITGB1, CDKN1A, RPLP0, HSP90B1, MT2A, TMSB4X, ...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>regulation of plasminogen activation (GO:0010755)</td>\n",
       "      <td>-2.122670</td>\n",
       "      <td>34.580925</td>\n",
       "      <td>4.934796e-06</td>\n",
       "      <td>8.410179e-08</td>\n",
       "      <td>[SERPINE2, CTSZ, SERPINE1, ENO1, THBS1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ribosome biogenesis (GO:0042254)</td>\n",
       "      <td>-1.202771</td>\n",
       "      <td>32.397044</td>\n",
       "      <td>2.411505e-10</td>\n",
       "      <td>2.005099e-12</td>\n",
       "      <td>[RPL4, NPM1, RPL41, RPL3, RPL10, RPS8, RPL23, ...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>cytoplasmic translation (GO:0002181)</td>\n",
       "      <td>-1.658984</td>\n",
       "      <td>30.246058</td>\n",
       "      <td>8.310536e-07</td>\n",
       "      <td>1.208048e-08</td>\n",
       "      <td>[RPL4, RPL41, RPS28, RPLP1, RPLP0, RPL36, RPS2...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>endodermal cell differentiation (GO:0035987)</td>\n",
       "      <td>-1.695379</td>\n",
       "      <td>27.695737</td>\n",
       "      <td>4.861713e-06</td>\n",
       "      <td>8.041931e-08</td>\n",
       "      <td>[COL4A2, COL11A1, MMP2, COL6A1, FN1, HMGA2, IT...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>negative regulation of protein processing (GO:...</td>\n",
       "      <td>-1.935167</td>\n",
       "      <td>27.269914</td>\n",
       "      <td>3.783623e-05</td>\n",
       "      <td>7.586212e-07</td>\n",
       "      <td>[PRNP, SERPINE2, CTSZ, SERPINE1, THBS1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>regulation of angiogenesis (GO:0045765)</td>\n",
       "      <td>-1.731712</td>\n",
       "      <td>27.188270</td>\n",
       "      <td>8.416245e-06</td>\n",
       "      <td>1.518721e-07</td>\n",
       "      <td>[ITGB1, BTG1, SPARC, SERPINE1, HMGA2, THBS1, F...</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1965</th>\n",
       "      <td>regulation of cardiac conduction (GO:1903779)</td>\n",
       "      <td>0.722729</td>\n",
       "      <td>-0.438432</td>\n",
       "      <td>5.850669e-01</td>\n",
       "      <td>5.451826e-01</td>\n",
       "      <td>[ATP1A1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1966</th>\n",
       "      <td>posttranscriptional regulation of gene express...</td>\n",
       "      <td>0.500688</td>\n",
       "      <td>-0.451175</td>\n",
       "      <td>4.691409e-01</td>\n",
       "      <td>4.061184e-01</td>\n",
       "      <td>[RPL10]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1967</th>\n",
       "      <td>RNA metabolic process (GO:0016070)</td>\n",
       "      <td>1.054473</td>\n",
       "      <td>-0.453473</td>\n",
       "      <td>6.737822e-01</td>\n",
       "      <td>6.504784e-01</td>\n",
       "      <td>[EIF4A1, HNRNPA1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1968</th>\n",
       "      <td>positive regulation of developmental growth (G...</td>\n",
       "      <td>0.583124</td>\n",
       "      <td>-0.462508</td>\n",
       "      <td>5.073458e-01</td>\n",
       "      <td>4.524151e-01</td>\n",
       "      <td>[PPIB]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1969</th>\n",
       "      <td>second-messenger-mediated signaling (GO:0019932)</td>\n",
       "      <td>0.727137</td>\n",
       "      <td>-0.478279</td>\n",
       "      <td>5.628728e-01</td>\n",
       "      <td>5.180122e-01</td>\n",
       "      <td>[PLCG2]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1970</th>\n",
       "      <td>protein deubiquitination (GO:0016579)</td>\n",
       "      <td>0.461414</td>\n",
       "      <td>-0.490693</td>\n",
       "      <td>4.228339e-01</td>\n",
       "      <td>3.452614e-01</td>\n",
       "      <td>[SMAD1, UBC, MDM2, ACTB]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1971</th>\n",
       "      <td>transcription initiation from RNA polymerase I...</td>\n",
       "      <td>0.406503</td>\n",
       "      <td>-0.513429</td>\n",
       "      <td>3.779766e-01</td>\n",
       "      <td>2.827937e-01</td>\n",
       "      <td>[WWTR1, CDKN1A, CTGF]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1972</th>\n",
       "      <td>positive regulation of peptidyl-serine phospho...</td>\n",
       "      <td>1.094811</td>\n",
       "      <td>-0.519321</td>\n",
       "      <td>6.486261e-01</td>\n",
       "      <td>6.222909e-01</td>\n",
       "      <td>[SPRY2]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1973</th>\n",
       "      <td>axonogenesis (GO:0007409)</td>\n",
       "      <td>0.383227</td>\n",
       "      <td>-0.519621</td>\n",
       "      <td>3.572911e-01</td>\n",
       "      <td>2.577127e-01</td>\n",
       "      <td>[RELN, LUM, S100A6, FAM129B]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1974</th>\n",
       "      <td>neuron differentiation (GO:0030182)</td>\n",
       "      <td>0.714479</td>\n",
       "      <td>-0.524243</td>\n",
       "      <td>5.315308e-01</td>\n",
       "      <td>4.801095e-01</td>\n",
       "      <td>[CTNNB1, SOX4]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1975</th>\n",
       "      <td>intracellular protein transport (GO:0006886)</td>\n",
       "      <td>0.533048</td>\n",
       "      <td>-0.527781</td>\n",
       "      <td>4.401469e-01</td>\n",
       "      <td>3.715325e-01</td>\n",
       "      <td>[EIF5A, PDIA3, CD74, NPM1, CALR]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1976</th>\n",
       "      <td>inositol phosphate metabolic process (GO:0043647)</td>\n",
       "      <td>0.629829</td>\n",
       "      <td>-0.536824</td>\n",
       "      <td>4.858398e-01</td>\n",
       "      <td>4.264188e-01</td>\n",
       "      <td>[PLCG2]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1977</th>\n",
       "      <td>protein oligomerization (GO:0051259)</td>\n",
       "      <td>0.691905</td>\n",
       "      <td>-0.538140</td>\n",
       "      <td>5.129075e-01</td>\n",
       "      <td>4.594314e-01</td>\n",
       "      <td>[PRNP, NPM1, VWF]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1978</th>\n",
       "      <td>positive regulation of response to DNA damage ...</td>\n",
       "      <td>0.662528</td>\n",
       "      <td>-0.544575</td>\n",
       "      <td>4.965663e-01</td>\n",
       "      <td>4.395670e-01</td>\n",
       "      <td>[HMGA2]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1979</th>\n",
       "      <td>negative regulation of autophagy (GO:0010507)</td>\n",
       "      <td>0.639648</td>\n",
       "      <td>-0.555306</td>\n",
       "      <td>4.806896e-01</td>\n",
       "      <td>4.197300e-01</td>\n",
       "      <td>[EIF4G2]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1980</th>\n",
       "      <td>regulation of cytokinesis (GO:0032465)</td>\n",
       "      <td>1.081651</td>\n",
       "      <td>-0.561527</td>\n",
       "      <td>6.237994e-01</td>\n",
       "      <td>5.950327e-01</td>\n",
       "      <td>[E2F7]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1981</th>\n",
       "      <td>regulation of hematopoietic progenitor cell di...</td>\n",
       "      <td>1.082614</td>\n",
       "      <td>-0.562027</td>\n",
       "      <td>6.237994e-01</td>\n",
       "      <td>5.950327e-01</td>\n",
       "      <td>[UBC]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1982</th>\n",
       "      <td>negative regulation of macromolecule metabolic...</td>\n",
       "      <td>0.655488</td>\n",
       "      <td>-0.564955</td>\n",
       "      <td>4.834298e-01</td>\n",
       "      <td>4.223650e-01</td>\n",
       "      <td>[CITED2, RACK1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1983</th>\n",
       "      <td>non-canonical Wnt signaling pathway (GO:0035567)</td>\n",
       "      <td>0.684568</td>\n",
       "      <td>-0.571126</td>\n",
       "      <td>4.921572e-01</td>\n",
       "      <td>4.341838e-01</td>\n",
       "      <td>[UBC, CTNNB1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1984</th>\n",
       "      <td>chloride transport (GO:0006821)</td>\n",
       "      <td>1.041566</td>\n",
       "      <td>-0.574973</td>\n",
       "      <td>6.087354e-01</td>\n",
       "      <td>5.757813e-01</td>\n",
       "      <td>[TTYH3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1985</th>\n",
       "      <td>positive regulation of cytoskeleton organizati...</td>\n",
       "      <td>1.114837</td>\n",
       "      <td>-0.578755</td>\n",
       "      <td>6.237994e-01</td>\n",
       "      <td>5.950327e-01</td>\n",
       "      <td>[CFL2]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1986</th>\n",
       "      <td>inflammatory response (GO:0006954)</td>\n",
       "      <td>0.532087</td>\n",
       "      <td>-0.586318</td>\n",
       "      <td>4.120391e-01</td>\n",
       "      <td>3.322321e-01</td>\n",
       "      <td>[SERPINA3, IL1B, PTX3, THBS1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1987</th>\n",
       "      <td>establishment of protein localization to organ...</td>\n",
       "      <td>0.896850</td>\n",
       "      <td>-0.589909</td>\n",
       "      <td>5.628728e-01</td>\n",
       "      <td>5.180122e-01</td>\n",
       "      <td>[PDIA3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1988</th>\n",
       "      <td>vesicle-mediated transport (GO:0016192)</td>\n",
       "      <td>0.935092</td>\n",
       "      <td>-0.625191</td>\n",
       "      <td>5.586349e-01</td>\n",
       "      <td>5.124320e-01</td>\n",
       "      <td>[HSPA8, ANXA2, APLP2, SYTL5, RHOB]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1989</th>\n",
       "      <td>negative regulation of neuron differentiation ...</td>\n",
       "      <td>0.989176</td>\n",
       "      <td>-0.629745</td>\n",
       "      <td>5.714634e-01</td>\n",
       "      <td>5.290691e-01</td>\n",
       "      <td>[CALR]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1990</th>\n",
       "      <td>regulation of cell differentiation (GO:0045595)</td>\n",
       "      <td>0.708207</td>\n",
       "      <td>-0.630709</td>\n",
       "      <td>4.738371e-01</td>\n",
       "      <td>4.104213e-01</td>\n",
       "      <td>[HSP90AB1, CTGF]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1991</th>\n",
       "      <td>glycosphingolipid metabolic process (GO:0006687)</td>\n",
       "      <td>0.852140</td>\n",
       "      <td>-0.641322</td>\n",
       "      <td>5.224698e-01</td>\n",
       "      <td>4.711394e-01</td>\n",
       "      <td>[PSAP]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1992</th>\n",
       "      <td>Golgi vesicle transport (GO:0048193)</td>\n",
       "      <td>0.732453</td>\n",
       "      <td>-0.705325</td>\n",
       "      <td>4.493282e-01</td>\n",
       "      <td>3.817600e-01</td>\n",
       "      <td>[CTSZ, TGFA, CD59, HYOU1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1993</th>\n",
       "      <td>positive regulation of protein serine/threonin...</td>\n",
       "      <td>0.795352</td>\n",
       "      <td>-0.732045</td>\n",
       "      <td>4.643685e-01</td>\n",
       "      <td>3.983585e-01</td>\n",
       "      <td>[CEMIP, IL1B]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1994</th>\n",
       "      <td>positive regulation of ERK1 and ERK2 cascade (...</td>\n",
       "      <td>1.044653</td>\n",
       "      <td>-0.928959</td>\n",
       "      <td>4.741885e-01</td>\n",
       "      <td>4.109634e-01</td>\n",
       "      <td>[CD74, SPRY2, CTGF]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1995 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              term_name    zscore  \\\n",
       "0                  protein targeting to ER (GO:0045047) -2.085357   \n",
       "1     SRP-dependent cotranslational protein targetin... -1.840410   \n",
       "2     nuclear-transcribed mRNA catabolic process, no... -1.997175   \n",
       "3        extracellular matrix organization (GO:0030198) -1.641114   \n",
       "4             peptide biosynthetic process (GO:0043043) -1.914538   \n",
       "5                      viral transcription (GO:0019083) -1.783768   \n",
       "6       cellular protein metabolic process (GO:0044267) -1.179625   \n",
       "7                     regulated exocytosis (GO:0045055) -2.282059   \n",
       "8     cotranslational protein targeting to membrane ... -1.368407   \n",
       "9                    viral gene expression (GO:0019080) -1.577981   \n",
       "10                           viral process (GO:0016032) -1.683366   \n",
       "11                  rRNA metabolic process (GO:0016072) -1.922755   \n",
       "12                        ncRNA processing (GO:0034470) -1.905479   \n",
       "13                  platelet degranulation (GO:0002576) -1.425928   \n",
       "14                         rRNA processing (GO:0006364) -1.601239   \n",
       "15    nuclear-transcribed mRNA catabolic process (GO... -1.227593   \n",
       "16    cellular macromolecule biosynthetic process (G... -1.736787   \n",
       "17                neutrophil degranulation (GO:0043312) -1.981217   \n",
       "18    antigen processing and presentation of exogeno... -2.294106   \n",
       "19         regulation of apoptotic process (GO:0042981) -1.759138   \n",
       "20            neutrophil mediated immunity (GO:0002446) -1.907256   \n",
       "21                             translation (GO:0006412) -1.219466   \n",
       "22    antigen processing and presentation of peptide... -1.728818   \n",
       "23     cytokine-mediated signaling pathway (GO:0019221) -1.331809   \n",
       "24    regulation of plasminogen activation (GO:0010755) -2.122670   \n",
       "25                     ribosome biogenesis (GO:0042254) -1.202771   \n",
       "26                 cytoplasmic translation (GO:0002181) -1.658984   \n",
       "27         endodermal cell differentiation (GO:0035987) -1.695379   \n",
       "28    negative regulation of protein processing (GO:... -1.935167   \n",
       "29              regulation of angiogenesis (GO:0045765) -1.731712   \n",
       "...                                                 ...       ...   \n",
       "1965      regulation of cardiac conduction (GO:1903779)  0.722729   \n",
       "1966  posttranscriptional regulation of gene express...  0.500688   \n",
       "1967                 RNA metabolic process (GO:0016070)  1.054473   \n",
       "1968  positive regulation of developmental growth (G...  0.583124   \n",
       "1969   second-messenger-mediated signaling (GO:0019932)  0.727137   \n",
       "1970              protein deubiquitination (GO:0016579)  0.461414   \n",
       "1971  transcription initiation from RNA polymerase I...  0.406503   \n",
       "1972  positive regulation of peptidyl-serine phospho...  1.094811   \n",
       "1973                          axonogenesis (GO:0007409)  0.383227   \n",
       "1974                neuron differentiation (GO:0030182)  0.714479   \n",
       "1975       intracellular protein transport (GO:0006886)  0.533048   \n",
       "1976  inositol phosphate metabolic process (GO:0043647)  0.629829   \n",
       "1977               protein oligomerization (GO:0051259)  0.691905   \n",
       "1978  positive regulation of response to DNA damage ...  0.662528   \n",
       "1979      negative regulation of autophagy (GO:0010507)  0.639648   \n",
       "1980             regulation of cytokinesis (GO:0032465)  1.081651   \n",
       "1981  regulation of hematopoietic progenitor cell di...  1.082614   \n",
       "1982  negative regulation of macromolecule metabolic...  0.655488   \n",
       "1983   non-canonical Wnt signaling pathway (GO:0035567)  0.684568   \n",
       "1984                    chloride transport (GO:0006821)  1.041566   \n",
       "1985  positive regulation of cytoskeleton organizati...  1.114837   \n",
       "1986                 inflammatory response (GO:0006954)  0.532087   \n",
       "1987  establishment of protein localization to organ...  0.896850   \n",
       "1988            vesicle-mediated transport (GO:0016192)  0.935092   \n",
       "1989  negative regulation of neuron differentiation ...  0.989176   \n",
       "1990    regulation of cell differentiation (GO:0045595)  0.708207   \n",
       "1991   glycosphingolipid metabolic process (GO:0006687)  0.852140   \n",
       "1992               Golgi vesicle transport (GO:0048193)  0.732453   \n",
       "1993  positive regulation of protein serine/threonin...  0.795352   \n",
       "1994  positive regulation of ERK1 and ERK2 cascade (...  1.044653   \n",
       "\n",
       "      combined_score           FDR        pvalue  \\\n",
       "0          91.031606  5.322236e-17  1.101143e-19   \n",
       "1          83.694450  1.773965e-17  1.778411e-20   \n",
       "2          81.207360  7.292814e-16  2.193327e-18   \n",
       "3          71.324472  5.322236e-17  1.333894e-19   \n",
       "4          70.881497  1.848915e-14  8.340968e-17   \n",
       "5          66.887449  1.293451e-14  5.186770e-17   \n",
       "6          65.464684  1.578545e-21  7.912506e-25   \n",
       "7          62.150413  2.118757e-10  1.486847e-12   \n",
       "8          60.951054  3.010519e-17  4.527096e-20   \n",
       "9          59.994607  8.770762e-15  3.077461e-17   \n",
       "10         53.959228  2.175298e-12  1.199413e-14   \n",
       "11         51.707043  2.411505e-10  2.093540e-12   \n",
       "12         51.169028  2.411505e-10  2.175794e-12   \n",
       "13         43.017045  1.214263e-11  7.912493e-14   \n",
       "14         42.777656  2.623370e-10  2.498448e-12   \n",
       "15         42.256861  2.240899e-13  1.123257e-15   \n",
       "16         41.183399  5.020480e-09  5.033063e-11   \n",
       "17         39.952382  1.515016e-07  1.746634e-09   \n",
       "18         38.742782  2.980108e-06  4.630745e-08   \n",
       "19         38.454118  3.049322e-08  3.209813e-10   \n",
       "20         37.839541  1.856361e-07  2.419317e-09   \n",
       "21         37.747814  5.990301e-12  3.603189e-14   \n",
       "22         35.926213  8.561212e-08  9.440935e-10   \n",
       "23         35.830235  2.411505e-10  2.070087e-12   \n",
       "24         34.580925  4.934796e-06  8.410179e-08   \n",
       "25         32.397044  2.411505e-10  2.005099e-12   \n",
       "26         30.246058  8.310536e-07  1.208048e-08   \n",
       "27         27.695737  4.861713e-06  8.041931e-08   \n",
       "28         27.269914  3.783623e-05  7.586212e-07   \n",
       "29         27.188270  8.416245e-06  1.518721e-07   \n",
       "...              ...           ...           ...   \n",
       "1965       -0.438432  5.850669e-01  5.451826e-01   \n",
       "1966       -0.451175  4.691409e-01  4.061184e-01   \n",
       "1967       -0.453473  6.737822e-01  6.504784e-01   \n",
       "1968       -0.462508  5.073458e-01  4.524151e-01   \n",
       "1969       -0.478279  5.628728e-01  5.180122e-01   \n",
       "1970       -0.490693  4.228339e-01  3.452614e-01   \n",
       "1971       -0.513429  3.779766e-01  2.827937e-01   \n",
       "1972       -0.519321  6.486261e-01  6.222909e-01   \n",
       "1973       -0.519621  3.572911e-01  2.577127e-01   \n",
       "1974       -0.524243  5.315308e-01  4.801095e-01   \n",
       "1975       -0.527781  4.401469e-01  3.715325e-01   \n",
       "1976       -0.536824  4.858398e-01  4.264188e-01   \n",
       "1977       -0.538140  5.129075e-01  4.594314e-01   \n",
       "1978       -0.544575  4.965663e-01  4.395670e-01   \n",
       "1979       -0.555306  4.806896e-01  4.197300e-01   \n",
       "1980       -0.561527  6.237994e-01  5.950327e-01   \n",
       "1981       -0.562027  6.237994e-01  5.950327e-01   \n",
       "1982       -0.564955  4.834298e-01  4.223650e-01   \n",
       "1983       -0.571126  4.921572e-01  4.341838e-01   \n",
       "1984       -0.574973  6.087354e-01  5.757813e-01   \n",
       "1985       -0.578755  6.237994e-01  5.950327e-01   \n",
       "1986       -0.586318  4.120391e-01  3.322321e-01   \n",
       "1987       -0.589909  5.628728e-01  5.180122e-01   \n",
       "1988       -0.625191  5.586349e-01  5.124320e-01   \n",
       "1989       -0.629745  5.714634e-01  5.290691e-01   \n",
       "1990       -0.630709  4.738371e-01  4.104213e-01   \n",
       "1991       -0.641322  5.224698e-01  4.711394e-01   \n",
       "1992       -0.705325  4.493282e-01  3.817600e-01   \n",
       "1993       -0.732045  4.643685e-01  3.983585e-01   \n",
       "1994       -0.928959  4.741885e-01  4.109634e-01   \n",
       "\n",
       "                                      overlapping_genes  \\\n",
       "0     [RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...   \n",
       "1     [RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...   \n",
       "2     [RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...   \n",
       "3     [ITGB1, SPARC, COL11A1, SERPINE1, TNC, LAMC1, ...   \n",
       "4     [RPL4, PCSK1, RPL41, RPL3, RPL10, RPS8, RPL23,...   \n",
       "5     [RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...   \n",
       "6     [RPL4, RPL3, RPL10, RPLP1, RPLP0, TNC, LAMC1, ...   \n",
       "7     [SERPINA3, SPARC, VWF, APLP2, SERPINE1, FN1, A...   \n",
       "8     [RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...   \n",
       "9     [RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...   \n",
       "10    [RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...   \n",
       "11    [RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...   \n",
       "12    [RPL4, SMAD1, RPL41, RPL3, RPL10, RPS8, RPL23,...   \n",
       "13    [SERPINA3, SPARC, VWF, APLP2, SERPINE1, FN1, A...   \n",
       "14    [RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...   \n",
       "15    [RPL4, EIF4A1, RPL41, RPL3, RPL10, RPS8, RPL23...   \n",
       "16    [RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...   \n",
       "17    [HSPA8, SERPINA3, HSP90AA1, HSP90AB1, ANXA2, T...   \n",
       "18                    [HLA-B, HLA-C, HLA-A, B2M, HLA-E]   \n",
       "19    [ITGB1, RPL10, CITED2, MAGED1, CLU, THBS1, FAM...   \n",
       "20    [HSPA8, SERPINA3, HSP90AA1, HSP90AB1, ANXA2, T...   \n",
       "21    [RPL4, RPL41, RPL3, RPL10, RPS8, RPL23, RPLP1,...   \n",
       "22    [PDIA3, CANX, HLA-B, HLA-C, CALR, HLA-A, B2M, ...   \n",
       "23    [ITGB1, CDKN1A, RPLP0, HSP90B1, MT2A, TMSB4X, ...   \n",
       "24              [SERPINE2, CTSZ, SERPINE1, ENO1, THBS1]   \n",
       "25    [RPL4, NPM1, RPL41, RPL3, RPL10, RPS8, RPL23, ...   \n",
       "26    [RPL4, RPL41, RPS28, RPLP1, RPLP0, RPL36, RPS2...   \n",
       "27    [COL4A2, COL11A1, MMP2, COL6A1, FN1, HMGA2, IT...   \n",
       "28              [PRNP, SERPINE2, CTSZ, SERPINE1, THBS1]   \n",
       "29    [ITGB1, BTG1, SPARC, SERPINE1, HMGA2, THBS1, F...   \n",
       "...                                                 ...   \n",
       "1965                                           [ATP1A1]   \n",
       "1966                                            [RPL10]   \n",
       "1967                                  [EIF4A1, HNRNPA1]   \n",
       "1968                                             [PPIB]   \n",
       "1969                                            [PLCG2]   \n",
       "1970                           [SMAD1, UBC, MDM2, ACTB]   \n",
       "1971                              [WWTR1, CDKN1A, CTGF]   \n",
       "1972                                            [SPRY2]   \n",
       "1973                       [RELN, LUM, S100A6, FAM129B]   \n",
       "1974                                     [CTNNB1, SOX4]   \n",
       "1975                   [EIF5A, PDIA3, CD74, NPM1, CALR]   \n",
       "1976                                            [PLCG2]   \n",
       "1977                                  [PRNP, NPM1, VWF]   \n",
       "1978                                            [HMGA2]   \n",
       "1979                                           [EIF4G2]   \n",
       "1980                                             [E2F7]   \n",
       "1981                                              [UBC]   \n",
       "1982                                    [CITED2, RACK1]   \n",
       "1983                                      [UBC, CTNNB1]   \n",
       "1984                                            [TTYH3]   \n",
       "1985                                             [CFL2]   \n",
       "1986                      [SERPINA3, IL1B, PTX3, THBS1]   \n",
       "1987                                            [PDIA3]   \n",
       "1988                 [HSPA8, ANXA2, APLP2, SYTL5, RHOB]   \n",
       "1989                                             [CALR]   \n",
       "1990                                   [HSP90AB1, CTGF]   \n",
       "1991                                             [PSAP]   \n",
       "1992                          [CTSZ, TGFA, CD59, HYOU1]   \n",
       "1993                                      [CEMIP, IL1B]   \n",
       "1994                                [CD74, SPRY2, CTGF]   \n",
       "\n",
       "                gene_set_library  \n",
       "0     GO_Biological_Process_2018  \n",
       "1     GO_Biological_Process_2018  \n",
       "2     GO_Biological_Process_2018  \n",
       "3     GO_Biological_Process_2018  \n",
       "4     GO_Biological_Process_2018  \n",
       "5     GO_Biological_Process_2018  \n",
       "6     GO_Biological_Process_2018  \n",
       "7     GO_Biological_Process_2018  \n",
       "8     GO_Biological_Process_2018  \n",
       "9     GO_Biological_Process_2018  \n",
       "10    GO_Biological_Process_2018  \n",
       "11    GO_Biological_Process_2018  \n",
       "12    GO_Biological_Process_2018  \n",
       "13    GO_Biological_Process_2018  \n",
       "14    GO_Biological_Process_2018  \n",
       "15    GO_Biological_Process_2018  \n",
       "16    GO_Biological_Process_2018  \n",
       "17    GO_Biological_Process_2018  \n",
       "18    GO_Biological_Process_2018  \n",
       "19    GO_Biological_Process_2018  \n",
       "20    GO_Biological_Process_2018  \n",
       "21    GO_Biological_Process_2018  \n",
       "22    GO_Biological_Process_2018  \n",
       "23    GO_Biological_Process_2018  \n",
       "24    GO_Biological_Process_2018  \n",
       "25    GO_Biological_Process_2018  \n",
       "26    GO_Biological_Process_2018  \n",
       "27    GO_Biological_Process_2018  \n",
       "28    GO_Biological_Process_2018  \n",
       "29    GO_Biological_Process_2018  \n",
       "...                          ...  \n",
       "1965  GO_Biological_Process_2018  \n",
       "1966  GO_Biological_Process_2018  \n",
       "1967  GO_Biological_Process_2018  \n",
       "1968  GO_Biological_Process_2018  \n",
       "1969  GO_Biological_Process_2018  \n",
       "1970  GO_Biological_Process_2018  \n",
       "1971  GO_Biological_Process_2018  \n",
       "1972  GO_Biological_Process_2018  \n",
       "1973  GO_Biological_Process_2018  \n",
       "1974  GO_Biological_Process_2018  \n",
       "1975  GO_Biological_Process_2018  \n",
       "1976  GO_Biological_Process_2018  \n",
       "1977  GO_Biological_Process_2018  \n",
       "1978  GO_Biological_Process_2018  \n",
       "1979  GO_Biological_Process_2018  \n",
       "1980  GO_Biological_Process_2018  \n",
       "1981  GO_Biological_Process_2018  \n",
       "1982  GO_Biological_Process_2018  \n",
       "1983  GO_Biological_Process_2018  \n",
       "1984  GO_Biological_Process_2018  \n",
       "1985  GO_Biological_Process_2018  \n",
       "1986  GO_Biological_Process_2018  \n",
       "1987  GO_Biological_Process_2018  \n",
       "1988  GO_Biological_Process_2018  \n",
       "1989  GO_Biological_Process_2018  \n",
       "1990  GO_Biological_Process_2018  \n",
       "1991  GO_Biological_Process_2018  \n",
       "1992  GO_Biological_Process_2018  \n",
       "1993  GO_Biological_Process_2018  \n",
       "1994  GO_Biological_Process_2018  \n",
       "\n",
       "[1995 rows x 7 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [],
   "source": [
    "c = Counter()\n",
    "for i in [enr_res[signature]['top'] for signature in enr_res.keys()]:\n",
    "    c.update(i['term_name'].iloc[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('protein targeting to ER (GO:0045047)', 17),\n",
       " ('SRP-dependent cotranslational protein targeting to membrane (GO:0006614)',\n",
       "  17),\n",
       " ('nuclear-transcribed mRNA catabolic process, nonsense-mediated decay (GO:0000184)',\n",
       "  17),\n",
       " ('viral transcription (GO:0019083)', 13),\n",
       " ('peptide biosynthetic process (GO:0043043)', 8),\n",
       " ('viral process (GO:0016032)', 5),\n",
       " ('neutrophil degranulation (GO:0043312)', 4),\n",
       " ('extracellular matrix organization (GO:0030198)', 3),\n",
       " ('neutrophil mediated immunity (GO:0002446)', 3),\n",
       " ('regulated exocytosis (GO:0045055)', 2),\n",
       " ('platelet degranulation (GO:0002576)', 2),\n",
       " ('rRNA metabolic process (GO:0016072)', 2),\n",
       " ('mitochondrial ATP synthesis coupled proton transport (GO:0042776)', 1),\n",
       " ('ribosomal small subunit export from nucleus (GO:0000056)', 1),\n",
       " ('positive regulation of nitric oxide biosynthetic process (GO:0045429)', 1),\n",
       " ('positive regulation of nitric oxide metabolic process (GO:1904407)', 1),\n",
       " ('ER to Golgi vesicle-mediated transport (GO:0006888)', 1),\n",
       " ('response to endoplasmic reticulum stress (GO:0034976)', 1),\n",
       " ('Golgi vesicle transport (GO:0048193)', 1),\n",
       " ('IRE1-mediated unfolded protein response (GO:0036498)', 1),\n",
       " ('vesicle coating (GO:0006901)', 1),\n",
       " ('response to unfolded protein (GO:0006986)', 1),\n",
       " ('retrograde protein transport, ER to cytosol (GO:0030970)', 1),\n",
       " ('endoplasmic reticulum to cytosol transport (GO:1903513)', 1),\n",
       " ('negative regulation of protein processing (GO:0010955)', 1),\n",
       " ('aorta development (GO:0035904)', 1),\n",
       " ('cotranslational protein targeting to membrane (GO:0006613)', 1),\n",
       " ('viral gene expression (GO:0019080)', 1),\n",
       " ('cellular protein metabolic process (GO:0044267)', 1)]"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c.most_common()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c = Counter()\n",
    "for i in [results[signature]['top'] for signature in results.keys()]\n",
    "for i in [results['1']['top']]: \n",
    "    c.update(i['Term name'].iloc[:5])\n",
    "[results[signature]['top'] for signature in results.keys()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a 3rd dictionary \n",
    "from collections import defaultdict\n",
    "results = defaultdict(dict)\n",
    "# Iterate through dictionary, create gene lists and save as 'genes'\n",
    "for signature in dict_top_genes:\n",
    "    for direction in ('top', 'bottom'):\n",
    "        genes = dict_top_genes[signature][direction] # value = dict [1st key] [2nd key]\n",
    "        results[signature][direction] = enrichr_result_to_df(genes,'','GO_Biological_Process_2018')\n",
    "        #for signature in list(dict_top_genes.keys())[0]:\n",
    "\n",
    "# freeze default dict for readonly1\n",
    "results.default_factory = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Number of studies to be processed\n",
    "n = len(study_names)\n",
    "i = 0\n",
    "for study in study_names[:]:\n",
    "    \n",
    "df_1_top=pd.DataFrame(results['1']['top'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if creating a list instead of dictionary\n",
    "test = []\n",
    "for signature in dict_top_genes:\n",
    "    for direction in ('top', 'bottom'):\n",
    "        genes = dict_top_genes[signature][direction] # value = dict [1st key] [2nd key]\n",
    "        genes_df = pd.DataFrame(enrichr_result_to_df(genes,'','GO_Biological_Process_2018'))\n",
    "        #specify names of each geneset (which study and top/bottom they are from)\n",
    "        test.append([signature, direction, genes_df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for study, d1 in D.items():\n",
    "    for top, v in d1.items():\n",
    "wanted_keys = dict_df.columns.get_values().tolist() # The keys you want\n",
    "dict((k, results[k]) for k in wanted_keys if k in results)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
