{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/maayanlab/Documents/projects/human-melanoma-signature\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/maayanlab/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.\n",
      "  from ._conv import register_converters as _register_converters\n"
     ]
    }
   ],
   "source": [
    "%cd /Users/maayanlab/Documents/projects/human-melanoma-signature\n",
    "import sys\n",
    "sys.path.append('pipeline/scripts/')\n",
    "import pandas as pd\n",
    "import h5py\n",
    "import numpy as np\n",
    "import json, operator, math, requests\n",
    "from geode_jupies import *\n",
    "import sklearn.metrics.pairwise as smp\n",
    "from scipy.spatial.distance import cosine\n",
    "import seaborn as sns; sns.set(color_codes=True)\n",
    "import RNAseq as RNA\n",
    "from collections import Counter\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/maayanlab/Documents/projects/human-melanoma-signature'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Melanoma import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Code Development for Melanoma Pipeline\n",
    "## Define input and output files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# infiles S4.4\n",
    "#infile = 's4-enrichment.dir/melanoma-enrichr-results.txt'\n",
    "#infile2 = 's4-enrichment.dir/enrichr-results-top.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# infiles\n",
    "infile = 's2-signatures.dir/melanoma-intersection-2000.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "#infiles S2.2\n",
    "#sig = 's2-signatures.dir/melanoma-signatures.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #infiles\n",
    "# cpm = 's1-expression_data.dir/melanoma-cpm.txt'\n",
    "# matches = 's2-expression_data.dir/melanoma-idmatches.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Infile & outfile\n",
    "# infiles = ['s1-expression_data.dir/melanoma-cpm.txt']\n",
    "# outfile = 's1-expression_data.dir/melanoma-IDstatus.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Infile & outfile\n",
    "# infiles = ['rawdata/human_matrix.h5', 'rawdata/m2.csv']\n",
    "# outfile = 's1-expression_data.dir/melanoma-cpm.txt'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDKN1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TNFRSF10B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STAT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IST1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PYCR1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene\n",
       "0     CDKN1A\n",
       "1  TNFRSF10B\n",
       "2      STAT3\n",
       "3       IST1\n",
       "4      PYCR1"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top2000_df = pd.read_table(infile)\n",
    "top2000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = top2000_df['gene'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df=run_enrichr(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term_name</th>\n",
       "      <th>zscore</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>FDR</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>overlapping_genes</th>\n",
       "      <th>gene_set_library</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>photoreceptor cell differentiation (GO:0046530)</td>\n",
       "      <td>-3.706326</td>\n",
       "      <td>23.530501</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.001749</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>astrocyte differentiation (GO:0048708)</td>\n",
       "      <td>-3.582280</td>\n",
       "      <td>22.742964</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.001749</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>leptin-mediated signaling pathway (GO:0033210)</td>\n",
       "      <td>-3.667175</td>\n",
       "      <td>22.361056</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002248</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T-helper 17 cell lineage commitment (GO:0072540)</td>\n",
       "      <td>-3.077399</td>\n",
       "      <td>19.537604</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.001749</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T-helper cell lineage commitment (GO:0002295)</td>\n",
       "      <td>-3.117394</td>\n",
       "      <td>19.008698</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002248</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cellular response to leptin stimulus (GO:0044320)</td>\n",
       "      <td>-2.982754</td>\n",
       "      <td>18.538732</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.001999</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>exit from host cell (GO:0035891)</td>\n",
       "      <td>-2.846877</td>\n",
       "      <td>18.074078</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.001749</td>\n",
       "      <td>[IST1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>negative regulation of response to reactive ox...</td>\n",
       "      <td>-2.928567</td>\n",
       "      <td>17.549039</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002498</td>\n",
       "      <td>[PYCR1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>interleukin-23-mediated signaling pathway (GO:...</td>\n",
       "      <td>-2.828802</td>\n",
       "      <td>16.951214</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002498</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>cellular response to growth hormone stimulus (...</td>\n",
       "      <td>-2.837055</td>\n",
       "      <td>16.730553</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002747</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>interleukin-21-mediated signaling pathway (GO:...</td>\n",
       "      <td>-2.733680</td>\n",
       "      <td>16.668955</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002248</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>regulation of hydrogen peroxide-induced cell d...</td>\n",
       "      <td>-2.830745</td>\n",
       "      <td>16.447317</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002997</td>\n",
       "      <td>[PYCR1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>radial glial cell differentiation (GO:0060019)</td>\n",
       "      <td>-2.549874</td>\n",
       "      <td>15.848251</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.001999</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>regulation of feeding behavior (GO:0060259)</td>\n",
       "      <td>-2.706660</td>\n",
       "      <td>15.726355</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002997</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>cellular response to UV-B (GO:0071493)</td>\n",
       "      <td>-2.568818</td>\n",
       "      <td>15.663690</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002248</td>\n",
       "      <td>[CDKN1A]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>stress-induced premature senescence (GO:0090400)</td>\n",
       "      <td>-2.525874</td>\n",
       "      <td>15.401832</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002248</td>\n",
       "      <td>[CDKN1A]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>response to leptin (GO:0044321)</td>\n",
       "      <td>-2.523836</td>\n",
       "      <td>15.389408</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002248</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>intrinsic apoptotic signaling pathway (GO:0097...</td>\n",
       "      <td>-1.776315</td>\n",
       "      <td>14.697694</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.000255</td>\n",
       "      <td>[CDKN1A, TNFRSF10B]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>miRNA mediated inhibition of translation (GO:0...</td>\n",
       "      <td>-2.421954</td>\n",
       "      <td>14.072140</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002997</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>viral release from host cell (GO:0019076)</td>\n",
       "      <td>-2.212909</td>\n",
       "      <td>14.049185</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.001749</td>\n",
       "      <td>[IST1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>cellular response to interleukin-21 (GO:0098757)</td>\n",
       "      <td>-2.291267</td>\n",
       "      <td>13.971285</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002248</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>eye photoreceptor cell differentiation (GO:000...</td>\n",
       "      <td>-2.427693</td>\n",
       "      <td>13.911410</td>\n",
       "      <td>0.018765</td>\n",
       "      <td>0.003246</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>negative regulation of hydrogen peroxide-induc...</td>\n",
       "      <td>-2.473317</td>\n",
       "      <td>13.819411</td>\n",
       "      <td>0.019098</td>\n",
       "      <td>0.003745</td>\n",
       "      <td>[PYCR1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ESCRT complex disassembly (GO:1904896)</td>\n",
       "      <td>-2.324270</td>\n",
       "      <td>13.706582</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002747</td>\n",
       "      <td>[IST1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>response to UV-B (GO:0010224)</td>\n",
       "      <td>-2.422943</td>\n",
       "      <td>13.704875</td>\n",
       "      <td>0.018765</td>\n",
       "      <td>0.003495</td>\n",
       "      <td>[CDKN1A]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ESCRT III complex disassembly (GO:1904903)</td>\n",
       "      <td>-2.313296</td>\n",
       "      <td>13.641863</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002747</td>\n",
       "      <td>[IST1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>cellular response to interleukin-15 (GO:0071350)</td>\n",
       "      <td>-2.393495</td>\n",
       "      <td>13.538310</td>\n",
       "      <td>0.018765</td>\n",
       "      <td>0.003495</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>activation of cysteine-type endopeptidase acti...</td>\n",
       "      <td>-2.317800</td>\n",
       "      <td>13.466984</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002997</td>\n",
       "      <td>[TNFRSF10B]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>glutamine family amino acid biosynthetic proce...</td>\n",
       "      <td>-2.353544</td>\n",
       "      <td>13.312334</td>\n",
       "      <td>0.018765</td>\n",
       "      <td>0.003495</td>\n",
       "      <td>[PYCR1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>interleukin-35-mediated signaling pathway (GO:...</td>\n",
       "      <td>-2.165554</td>\n",
       "      <td>12.582396</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.002997</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>positive regulation of nucleic acid-templated ...</td>\n",
       "      <td>-1.642398</td>\n",
       "      <td>3.487894</td>\n",
       "      <td>0.125113</td>\n",
       "      <td>0.119593</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>cellular response to tumor necrosis factor (GO...</td>\n",
       "      <td>-1.145458</td>\n",
       "      <td>3.482709</td>\n",
       "      <td>0.059115</td>\n",
       "      <td>0.047813</td>\n",
       "      <td>[TNFRSF10B]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>regulation of I-kappaB kinase/NF-kappaB signal...</td>\n",
       "      <td>-1.156602</td>\n",
       "      <td>3.459908</td>\n",
       "      <td>0.060975</td>\n",
       "      <td>0.050215</td>\n",
       "      <td>[TNFRSF10B]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>Ras protein signal transduction (GO:0007265)</td>\n",
       "      <td>-1.187258</td>\n",
       "      <td>3.448635</td>\n",
       "      <td>0.063840</td>\n",
       "      <td>0.054765</td>\n",
       "      <td>[CDKN1A]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>transmembrane receptor protein tyrosine kinase...</td>\n",
       "      <td>-1.465393</td>\n",
       "      <td>3.443255</td>\n",
       "      <td>0.104628</td>\n",
       "      <td>0.095396</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>regulation of cell cycle (GO:0051726)</td>\n",
       "      <td>-1.163793</td>\n",
       "      <td>3.421870</td>\n",
       "      <td>0.063422</td>\n",
       "      <td>0.052851</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>regulation of nucleic acid-templated transcrip...</td>\n",
       "      <td>-1.737924</td>\n",
       "      <td>3.379517</td>\n",
       "      <td>0.147383</td>\n",
       "      <td>0.143049</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>phosphorylation (GO:0016310)</td>\n",
       "      <td>-1.420382</td>\n",
       "      <td>3.372309</td>\n",
       "      <td>0.102647</td>\n",
       "      <td>0.093086</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>positive regulation of signal transduction (GO...</td>\n",
       "      <td>-1.089970</td>\n",
       "      <td>3.250216</td>\n",
       "      <td>0.061194</td>\n",
       "      <td>0.050695</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>activation of protein kinase activity (GO:0032...</td>\n",
       "      <td>-1.121660</td>\n",
       "      <td>3.210225</td>\n",
       "      <td>0.065871</td>\n",
       "      <td>0.057153</td>\n",
       "      <td>[TNFRSF10B]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>regulation of cell growth (GO:0001558)</td>\n",
       "      <td>-1.045833</td>\n",
       "      <td>3.143651</td>\n",
       "      <td>0.060825</td>\n",
       "      <td>0.049495</td>\n",
       "      <td>[CDKN1A]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>mitotic cell cycle phase transition (GO:0044772)</td>\n",
       "      <td>-1.044493</td>\n",
       "      <td>3.043102</td>\n",
       "      <td>0.063647</td>\n",
       "      <td>0.054287</td>\n",
       "      <td>[CDKN1A]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>positive regulation of cell proliferation (GO:...</td>\n",
       "      <td>-1.322691</td>\n",
       "      <td>3.021504</td>\n",
       "      <td>0.110506</td>\n",
       "      <td>0.101839</td>\n",
       "      <td>[CDKN1A]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>cellular response to DNA damage stimulus (GO:0...</td>\n",
       "      <td>-1.193166</td>\n",
       "      <td>3.016154</td>\n",
       "      <td>0.089974</td>\n",
       "      <td>0.079830</td>\n",
       "      <td>[CDKN1A]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>regulation of programmed cell death (GO:0043067)</td>\n",
       "      <td>-1.074215</td>\n",
       "      <td>2.928459</td>\n",
       "      <td>0.075035</td>\n",
       "      <td>0.065472</td>\n",
       "      <td>[TNFRSF10B]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>positive regulation of intracellular signal tr...</td>\n",
       "      <td>-1.327695</td>\n",
       "      <td>2.878660</td>\n",
       "      <td>0.122173</td>\n",
       "      <td>0.114387</td>\n",
       "      <td>[TNFRSF10B]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>positive regulation of MAPK cascade (GO:0043410)</td>\n",
       "      <td>-0.980793</td>\n",
       "      <td>2.602114</td>\n",
       "      <td>0.079826</td>\n",
       "      <td>0.070435</td>\n",
       "      <td>[TNFRSF10B]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>negative regulation of programmed cell death (...</td>\n",
       "      <td>-1.107159</td>\n",
       "      <td>2.569865</td>\n",
       "      <td>0.107086</td>\n",
       "      <td>0.098162</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>positive regulation of transcription from RNA ...</td>\n",
       "      <td>-1.565668</td>\n",
       "      <td>2.559511</td>\n",
       "      <td>0.196927</td>\n",
       "      <td>0.194997</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>nervous system development (GO:0007399)</td>\n",
       "      <td>-1.144144</td>\n",
       "      <td>2.536633</td>\n",
       "      <td>0.117574</td>\n",
       "      <td>0.108929</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>organelle assembly (GO:0070925)</td>\n",
       "      <td>-1.052614</td>\n",
       "      <td>2.525647</td>\n",
       "      <td>0.100638</td>\n",
       "      <td>0.090772</td>\n",
       "      <td>[IST1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>negative regulation of cellular process (GO:00...</td>\n",
       "      <td>-1.191996</td>\n",
       "      <td>2.461689</td>\n",
       "      <td>0.131971</td>\n",
       "      <td>0.126795</td>\n",
       "      <td>[CDKN1A]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>negative regulation of transcription, DNA-temp...</td>\n",
       "      <td>-1.431437</td>\n",
       "      <td>2.395330</td>\n",
       "      <td>0.190413</td>\n",
       "      <td>0.187613</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>positive regulation of gene expression (GO:001...</td>\n",
       "      <td>-1.388211</td>\n",
       "      <td>2.390713</td>\n",
       "      <td>0.182255</td>\n",
       "      <td>0.178681</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>neutrophil activation involved in immune respo...</td>\n",
       "      <td>-1.089808</td>\n",
       "      <td>2.354274</td>\n",
       "      <td>0.122191</td>\n",
       "      <td>0.115295</td>\n",
       "      <td>[IST1]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>transcription from RNA polymerase II promoter ...</td>\n",
       "      <td>-1.043357</td>\n",
       "      <td>2.249833</td>\n",
       "      <td>0.122191</td>\n",
       "      <td>0.115748</td>\n",
       "      <td>[CDKN1A]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>regulation of gene expression (GO:0010468)</td>\n",
       "      <td>-1.540156</td>\n",
       "      <td>2.237325</td>\n",
       "      <td>0.235099</td>\n",
       "      <td>0.233947</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>regulation of cellular macromolecule biosynthe...</td>\n",
       "      <td>-1.146952</td>\n",
       "      <td>2.188678</td>\n",
       "      <td>0.152066</td>\n",
       "      <td>0.148339</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>positive regulation of transcription, DNA-temp...</td>\n",
       "      <td>-1.386738</td>\n",
       "      <td>1.919279</td>\n",
       "      <td>0.250568</td>\n",
       "      <td>0.250568</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>negative regulation of transcription from RNA ...</td>\n",
       "      <td>-0.931151</td>\n",
       "      <td>1.873433</td>\n",
       "      <td>0.138479</td>\n",
       "      <td>0.133727</td>\n",
       "      <td>[STAT3]</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>204 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             term_name    zscore  \\\n",
       "0      photoreceptor cell differentiation (GO:0046530) -3.706326   \n",
       "1               astrocyte differentiation (GO:0048708) -3.582280   \n",
       "2       leptin-mediated signaling pathway (GO:0033210) -3.667175   \n",
       "3     T-helper 17 cell lineage commitment (GO:0072540) -3.077399   \n",
       "4        T-helper cell lineage commitment (GO:0002295) -3.117394   \n",
       "5    cellular response to leptin stimulus (GO:0044320) -2.982754   \n",
       "6                     exit from host cell (GO:0035891) -2.846877   \n",
       "7    negative regulation of response to reactive ox... -2.928567   \n",
       "8    interleukin-23-mediated signaling pathway (GO:... -2.828802   \n",
       "9    cellular response to growth hormone stimulus (... -2.837055   \n",
       "10   interleukin-21-mediated signaling pathway (GO:... -2.733680   \n",
       "11   regulation of hydrogen peroxide-induced cell d... -2.830745   \n",
       "12      radial glial cell differentiation (GO:0060019) -2.549874   \n",
       "13         regulation of feeding behavior (GO:0060259) -2.706660   \n",
       "14              cellular response to UV-B (GO:0071493) -2.568818   \n",
       "15    stress-induced premature senescence (GO:0090400) -2.525874   \n",
       "16                     response to leptin (GO:0044321) -2.523836   \n",
       "17   intrinsic apoptotic signaling pathway (GO:0097... -1.776315   \n",
       "18   miRNA mediated inhibition of translation (GO:0... -2.421954   \n",
       "19           viral release from host cell (GO:0019076) -2.212909   \n",
       "20    cellular response to interleukin-21 (GO:0098757) -2.291267   \n",
       "21   eye photoreceptor cell differentiation (GO:000... -2.427693   \n",
       "22   negative regulation of hydrogen peroxide-induc... -2.473317   \n",
       "23              ESCRT complex disassembly (GO:1904896) -2.324270   \n",
       "24                       response to UV-B (GO:0010224) -2.422943   \n",
       "25          ESCRT III complex disassembly (GO:1904903) -2.313296   \n",
       "26    cellular response to interleukin-15 (GO:0071350) -2.393495   \n",
       "27   activation of cysteine-type endopeptidase acti... -2.317800   \n",
       "28   glutamine family amino acid biosynthetic proce... -2.353544   \n",
       "29   interleukin-35-mediated signaling pathway (GO:... -2.165554   \n",
       "..                                                 ...       ...   \n",
       "174  positive regulation of nucleic acid-templated ... -1.642398   \n",
       "175  cellular response to tumor necrosis factor (GO... -1.145458   \n",
       "176  regulation of I-kappaB kinase/NF-kappaB signal... -1.156602   \n",
       "177       Ras protein signal transduction (GO:0007265) -1.187258   \n",
       "178  transmembrane receptor protein tyrosine kinase... -1.465393   \n",
       "179              regulation of cell cycle (GO:0051726) -1.163793   \n",
       "180  regulation of nucleic acid-templated transcrip... -1.737924   \n",
       "181                       phosphorylation (GO:0016310) -1.420382   \n",
       "182  positive regulation of signal transduction (GO... -1.089970   \n",
       "183  activation of protein kinase activity (GO:0032... -1.121660   \n",
       "184             regulation of cell growth (GO:0001558) -1.045833   \n",
       "185   mitotic cell cycle phase transition (GO:0044772) -1.044493   \n",
       "186  positive regulation of cell proliferation (GO:... -1.322691   \n",
       "187  cellular response to DNA damage stimulus (GO:0... -1.193166   \n",
       "188   regulation of programmed cell death (GO:0043067) -1.074215   \n",
       "189  positive regulation of intracellular signal tr... -1.327695   \n",
       "190   positive regulation of MAPK cascade (GO:0043410) -0.980793   \n",
       "191  negative regulation of programmed cell death (... -1.107159   \n",
       "192  positive regulation of transcription from RNA ... -1.565668   \n",
       "193            nervous system development (GO:0007399) -1.144144   \n",
       "194                    organelle assembly (GO:0070925) -1.052614   \n",
       "195  negative regulation of cellular process (GO:00... -1.191996   \n",
       "196  negative regulation of transcription, DNA-temp... -1.431437   \n",
       "197  positive regulation of gene expression (GO:001... -1.388211   \n",
       "198  neutrophil activation involved in immune respo... -1.089808   \n",
       "199  transcription from RNA polymerase II promoter ... -1.043357   \n",
       "200         regulation of gene expression (GO:0010468) -1.540156   \n",
       "201  regulation of cellular macromolecule biosynthe... -1.146952   \n",
       "202  positive regulation of transcription, DNA-temp... -1.386738   \n",
       "203  negative regulation of transcription from RNA ... -0.931151   \n",
       "\n",
       "     combined_score       FDR    pvalue    overlapping_genes  \\\n",
       "0         23.530501  0.018525  0.001749              [STAT3]   \n",
       "1         22.742964  0.018525  0.001749              [STAT3]   \n",
       "2         22.361056  0.018525  0.002248              [STAT3]   \n",
       "3         19.537604  0.018525  0.001749              [STAT3]   \n",
       "4         19.008698  0.018525  0.002248              [STAT3]   \n",
       "5         18.538732  0.018525  0.001999              [STAT3]   \n",
       "6         18.074078  0.018525  0.001749               [IST1]   \n",
       "7         17.549039  0.018525  0.002498              [PYCR1]   \n",
       "8         16.951214  0.018525  0.002498              [STAT3]   \n",
       "9         16.730553  0.018525  0.002747              [STAT3]   \n",
       "10        16.668955  0.018525  0.002248              [STAT3]   \n",
       "11        16.447317  0.018525  0.002997              [PYCR1]   \n",
       "12        15.848251  0.018525  0.001999              [STAT3]   \n",
       "13        15.726355  0.018525  0.002997              [STAT3]   \n",
       "14        15.663690  0.018525  0.002248             [CDKN1A]   \n",
       "15        15.401832  0.018525  0.002248             [CDKN1A]   \n",
       "16        15.389408  0.018525  0.002248              [STAT3]   \n",
       "17        14.697694  0.018525  0.000255  [CDKN1A, TNFRSF10B]   \n",
       "18        14.072140  0.018525  0.002997              [STAT3]   \n",
       "19        14.049185  0.018525  0.001749               [IST1]   \n",
       "20        13.971285  0.018525  0.002248              [STAT3]   \n",
       "21        13.911410  0.018765  0.003246              [STAT3]   \n",
       "22        13.819411  0.019098  0.003745              [PYCR1]   \n",
       "23        13.706582  0.018525  0.002747               [IST1]   \n",
       "24        13.704875  0.018765  0.003495             [CDKN1A]   \n",
       "25        13.641863  0.018525  0.002747               [IST1]   \n",
       "26        13.538310  0.018765  0.003495              [STAT3]   \n",
       "27        13.466984  0.018525  0.002997          [TNFRSF10B]   \n",
       "28        13.312334  0.018765  0.003495              [PYCR1]   \n",
       "29        12.582396  0.018525  0.002997              [STAT3]   \n",
       "..              ...       ...       ...                  ...   \n",
       "174        3.487894  0.125113  0.119593              [STAT3]   \n",
       "175        3.482709  0.059115  0.047813          [TNFRSF10B]   \n",
       "176        3.459908  0.060975  0.050215          [TNFRSF10B]   \n",
       "177        3.448635  0.063840  0.054765             [CDKN1A]   \n",
       "178        3.443255  0.104628  0.095396              [STAT3]   \n",
       "179        3.421870  0.063422  0.052851              [STAT3]   \n",
       "180        3.379517  0.147383  0.143049              [STAT3]   \n",
       "181        3.372309  0.102647  0.093086              [STAT3]   \n",
       "182        3.250216  0.061194  0.050695              [STAT3]   \n",
       "183        3.210225  0.065871  0.057153          [TNFRSF10B]   \n",
       "184        3.143651  0.060825  0.049495             [CDKN1A]   \n",
       "185        3.043102  0.063647  0.054287             [CDKN1A]   \n",
       "186        3.021504  0.110506  0.101839             [CDKN1A]   \n",
       "187        3.016154  0.089974  0.079830             [CDKN1A]   \n",
       "188        2.928459  0.075035  0.065472          [TNFRSF10B]   \n",
       "189        2.878660  0.122173  0.114387          [TNFRSF10B]   \n",
       "190        2.602114  0.079826  0.070435          [TNFRSF10B]   \n",
       "191        2.569865  0.107086  0.098162              [STAT3]   \n",
       "192        2.559511  0.196927  0.194997              [STAT3]   \n",
       "193        2.536633  0.117574  0.108929              [STAT3]   \n",
       "194        2.525647  0.100638  0.090772               [IST1]   \n",
       "195        2.461689  0.131971  0.126795             [CDKN1A]   \n",
       "196        2.395330  0.190413  0.187613              [STAT3]   \n",
       "197        2.390713  0.182255  0.178681              [STAT3]   \n",
       "198        2.354274  0.122191  0.115295               [IST1]   \n",
       "199        2.249833  0.122191  0.115748             [CDKN1A]   \n",
       "200        2.237325  0.235099  0.233947              [STAT3]   \n",
       "201        2.188678  0.152066  0.148339              [STAT3]   \n",
       "202        1.919279  0.250568  0.250568              [STAT3]   \n",
       "203        1.873433  0.138479  0.133727              [STAT3]   \n",
       "\n",
       "               gene_set_library  \n",
       "0    GO_Biological_Process_2018  \n",
       "1    GO_Biological_Process_2018  \n",
       "2    GO_Biological_Process_2018  \n",
       "3    GO_Biological_Process_2018  \n",
       "4    GO_Biological_Process_2018  \n",
       "5    GO_Biological_Process_2018  \n",
       "6    GO_Biological_Process_2018  \n",
       "7    GO_Biological_Process_2018  \n",
       "8    GO_Biological_Process_2018  \n",
       "9    GO_Biological_Process_2018  \n",
       "10   GO_Biological_Process_2018  \n",
       "11   GO_Biological_Process_2018  \n",
       "12   GO_Biological_Process_2018  \n",
       "13   GO_Biological_Process_2018  \n",
       "14   GO_Biological_Process_2018  \n",
       "15   GO_Biological_Process_2018  \n",
       "16   GO_Biological_Process_2018  \n",
       "17   GO_Biological_Process_2018  \n",
       "18   GO_Biological_Process_2018  \n",
       "19   GO_Biological_Process_2018  \n",
       "20   GO_Biological_Process_2018  \n",
       "21   GO_Biological_Process_2018  \n",
       "22   GO_Biological_Process_2018  \n",
       "23   GO_Biological_Process_2018  \n",
       "24   GO_Biological_Process_2018  \n",
       "25   GO_Biological_Process_2018  \n",
       "26   GO_Biological_Process_2018  \n",
       "27   GO_Biological_Process_2018  \n",
       "28   GO_Biological_Process_2018  \n",
       "29   GO_Biological_Process_2018  \n",
       "..                          ...  \n",
       "174  GO_Biological_Process_2018  \n",
       "175  GO_Biological_Process_2018  \n",
       "176  GO_Biological_Process_2018  \n",
       "177  GO_Biological_Process_2018  \n",
       "178  GO_Biological_Process_2018  \n",
       "179  GO_Biological_Process_2018  \n",
       "180  GO_Biological_Process_2018  \n",
       "181  GO_Biological_Process_2018  \n",
       "182  GO_Biological_Process_2018  \n",
       "183  GO_Biological_Process_2018  \n",
       "184  GO_Biological_Process_2018  \n",
       "185  GO_Biological_Process_2018  \n",
       "186  GO_Biological_Process_2018  \n",
       "187  GO_Biological_Process_2018  \n",
       "188  GO_Biological_Process_2018  \n",
       "189  GO_Biological_Process_2018  \n",
       "190  GO_Biological_Process_2018  \n",
       "191  GO_Biological_Process_2018  \n",
       "192  GO_Biological_Process_2018  \n",
       "193  GO_Biological_Process_2018  \n",
       "194  GO_Biological_Process_2018  \n",
       "195  GO_Biological_Process_2018  \n",
       "196  GO_Biological_Process_2018  \n",
       "197  GO_Biological_Process_2018  \n",
       "198  GO_Biological_Process_2018  \n",
       "199  GO_Biological_Process_2018  \n",
       "200  GO_Biological_Process_2018  \n",
       "201  GO_Biological_Process_2018  \n",
       "202  GO_Biological_Process_2018  \n",
       "203  GO_Biological_Process_2018  \n",
       "\n",
       "[204 rows x 7 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Select rows whose column value equals a scalar 'top'\n",
    "\n",
    "for direction in ['top', 'bottom']:\n",
    "    all_top = enr_res_df.loc[enr_res_df['direction'] == direction]\n",
    "\n",
    "    # Create a table of pathways to study names, filtered by FDR value\n",
    "    # Fill na value is set to 1 since that'd be the lowest for log transformation later\n",
    "    toptable= all_top.pivot(index='term_name', columns = 'signature',values='FDR').fillna(1)\n",
    "    # Log 10 transformation\n",
    "    all_top_log = -np.log10(toptable)\n",
    "    outfile = '{outfileRoot}{direction}.txt'.format(**locals())\n",
    "    all_top_log.to_csv (outfile, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [],
   "source": [
    "c = Counter()\n",
    "for i in [enr_res[signature]['top'] for signature in enr_res.keys()]:\n",
    "    c.update(i['term_name'].iloc[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('protein targeting to ER (GO:0045047)', 17),\n",
       " ('SRP-dependent cotranslational protein targeting to membrane (GO:0006614)',\n",
       "  17),\n",
       " ('nuclear-transcribed mRNA catabolic process, nonsense-mediated decay (GO:0000184)',\n",
       "  17),\n",
       " ('viral transcription (GO:0019083)', 13),\n",
       " ('peptide biosynthetic process (GO:0043043)', 8),\n",
       " ('viral process (GO:0016032)', 5),\n",
       " ('neutrophil degranulation (GO:0043312)', 4),\n",
       " ('extracellular matrix organization (GO:0030198)', 3),\n",
       " ('neutrophil mediated immunity (GO:0002446)', 3),\n",
       " ('regulated exocytosis (GO:0045055)', 2),\n",
       " ('platelet degranulation (GO:0002576)', 2),\n",
       " ('rRNA metabolic process (GO:0016072)', 2),\n",
       " ('mitochondrial ATP synthesis coupled proton transport (GO:0042776)', 1),\n",
       " ('ribosomal small subunit export from nucleus (GO:0000056)', 1),\n",
       " ('positive regulation of nitric oxide biosynthetic process (GO:0045429)', 1),\n",
       " ('positive regulation of nitric oxide metabolic process (GO:1904407)', 1),\n",
       " ('ER to Golgi vesicle-mediated transport (GO:0006888)', 1),\n",
       " ('response to endoplasmic reticulum stress (GO:0034976)', 1),\n",
       " ('Golgi vesicle transport (GO:0048193)', 1),\n",
       " ('IRE1-mediated unfolded protein response (GO:0036498)', 1),\n",
       " ('vesicle coating (GO:0006901)', 1),\n",
       " ('response to unfolded protein (GO:0006986)', 1),\n",
       " ('retrograde protein transport, ER to cytosol (GO:0030970)', 1),\n",
       " ('endoplasmic reticulum to cytosol transport (GO:1903513)', 1),\n",
       " ('negative regulation of protein processing (GO:0010955)', 1),\n",
       " ('aorta development (GO:0035904)', 1),\n",
       " ('cotranslational protein targeting to membrane (GO:0006613)', 1),\n",
       " ('viral gene expression (GO:0019080)', 1),\n",
       " ('cellular protein metabolic process (GO:0044267)', 1)]"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c.most_common()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c = Counter()\n",
    "for i in [results[signature]['top'] for signature in results.keys()]\n",
    "for i in [results['1']['top']]: \n",
    "    c.update(i['Term name'].iloc[:5])\n",
    "[results[signature]['top'] for signature in results.keys()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a 3rd dictionary \n",
    "from collections import defaultdict\n",
    "results = defaultdict(dict)\n",
    "# Iterate through dictionary, create gene lists and save as 'genes'\n",
    "for signature in dict_top_genes:\n",
    "    for direction in ('top', 'bottom'):\n",
    "        genes = dict_top_genes[signature][direction] # value = dict [1st key] [2nd key]\n",
    "        results[signature][direction] = enrichr_result_to_df(genes,'','GO_Biological_Process_2018')\n",
    "        #for signature in list(dict_top_genes.keys())[0]:\n",
    "\n",
    "# freeze default dict for readonly1\n",
    "results.default_factory = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Number of studies to be processed\n",
    "n = len(study_names)\n",
    "i = 0\n",
    "for study in study_names[:]:\n",
    "    \n",
    "df_1_top=pd.DataFrame(results['1']['top'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if creating a list instead of dictionary\n",
    "test = []\n",
    "for signature in dict_top_genes:\n",
    "    for direction in ('top', 'bottom'):\n",
    "        genes = dict_top_genes[signature][direction] # value = dict [1st key] [2nd key]\n",
    "        genes_df = pd.DataFrame(enrichr_result_to_df(genes,'','GO_Biological_Process_2018'))\n",
    "        #specify names of each geneset (which study and top/bottom they are from)\n",
    "        test.append([signature, direction, genes_df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for study, d1 in D.items():\n",
    "    for top, v in d1.items():\n",
    "wanted_keys = dict_df.columns.get_values().tolist() # The keys you want\n",
    "dict((k, results[k]) for k in wanted_keys if k in results)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
